File: csharp/CsSearch/CsSearch/FileTypes.cs
﻿using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Xml.Linq;

namespace CsSearch
{
	public enum FileType
	{
		Archive,
		Binary,
		Text,
		Unknown
	};

	public class FileTypes
	{
		public readonly ISet<string> CurrentAndParentDirs = new HashSet<string> {".", ".."};

		private readonly string _fileTypesResource;
		private readonly IDictionary<string, ISet<string>> _fileTypesDictionary;

		public FileTypes()
		{
			_fileTypesResource = Properties.Resources.filetypes;
			_fileTypesDictionary = new Dictionary<string, ISet<string>>();
			PopulateFileTypes();
		}

		private void PopulateFileTypes()
		{
			var doc = XDocument.Parse(_fileTypesResource);
			foreach (var f in doc.Descendants("filetype"))
			{
				var name = f.Attributes("name").First().Value;
				var extensions = f.Descendants("extensions").First().Value;
				var extensionSet = new HashSet<string>(extensions.Split(new[]{' ','\n'}).Select(x => "." + x));
				_fileTypesDictionary[name] = extensionSet;
			}
			_fileTypesDictionary["text"].UnionWith(_fileTypesDictionary["code"]);
			_fileTypesDictionary["text"].UnionWith(_fileTypesDictionary["xml"]);
			_fileTypesDictionary["searchable"] = new HashSet<string>(_fileTypesDictionary["text"]);
			_fileTypesDictionary["searchable"].UnionWith(_fileTypesDictionary["binary"]);
			_fileTypesDictionary["searchable"].UnionWith(_fileTypesDictionary["archive"]);
		}

		public FileType GetFileType(FileInfo f)
		{
			if (IsArchiveFile(f)) return FileType.Archive;
			if (IsBinaryFile(f)) return FileType.Binary;
			return IsTextFile(f) ? FileType.Text : FileType.Unknown;
		}

		public bool IsBinaryFile(FileInfo f)
		{
			return _fileTypesDictionary["binary"].Contains(f.Extension.ToLowerInvariant());
		}

		public bool IsArchiveFile(FileInfo f)
		{
			return _fileTypesDictionary["archive"].Contains(f.Extension.ToLowerInvariant());
		}

		public bool IsSearchableFile(FileInfo f)
		{
			return _fileTypesDictionary["searchable"].Contains(f.Extension.ToLowerInvariant());
		}

		public bool IsTextFile(FileInfo f)
		{
			return _fileTypesDictionary["text"].Contains(f.Extension.ToLowerInvariant());
		}

		public bool IsUnknownFile(FileInfo f)
		{
			return !IsSearchableFile(f);
		}
	}
}

File: ruby/rbsearch/filetypes.rb
################################################################################
#
# filetypes.rb
#
# Provides information on file type (archive, binary, text, unknown)
#
################################################################################
require 'rexml/document'
require 'fileutil.rb'
include REXML

module FileType
  Archive = 1
  Binary  = 2
  Text    = 3
  Unknown = 4
end

class FileTypes
  def initialize
    # TODO: move to config
    @file_types_path = '~/src/git/xsearch/shared/filetypes.xml'
    set_file_type_map
  end

  def set_file_type_map
    @file_type_map = {}
    doc = Document.new(File.new(File.expand_path(@file_types_path)))
    doc.elements.each('filetypes/filetype') { |filetype|
      name = filetype.attributes['name']
      filetype.elements.each('extensions') { |extensions|
        exts = extensions.text.split(' ')
        @file_type_map[name] = exts.to_set
      }
    }
    @file_type_map['text'] = @file_type_map['text'] + @file_type_map['code'] +
      @file_type_map['xml']
    @file_type_map['searchable'] = @file_type_map['text'] +
      @file_type_map['archive'] + @file_type_map['binary']
  end

  def get_filetype(filename)
    if is_text_file(filename)
      FileType::Text
    elsif is_binary_file(filename)
      FileType::Binary
    elsif is_archive_file(filename)
      FileType::Archive
    else
      FileType::Unknown
    end
  end

  def is_archive_file(f)
    @file_type_map['archive'].include?(FileUtil::get_extension(f))
  end

  def is_binary_file(f)
    @file_type_map['binary'].include?(FileUtil::get_extension(f))
  end

  def is_searchable_file(f)
    @file_type_map['searchable'].include?(FileUtil::get_extension(f))
  end

  def is_text_file(f)
    @file_type_map['text'].include?(FileUtil::get_extension(f))
  end
end
